src/transactions/TransactionFrame.cpp

```diff
@@ -1329,6 +1334,143 @@ TransactionFrame::applyOperations(SignatureChecker& signatureChecker,
     return false;
 }
 
+bool
+TransactionFrame::applyLifetimeBumps(Application& app, AbstractLedgerTxn& ltx)
+{
+    if (!isSoroban())
+    {
+        return true;
+    }
+
+#ifdef ENABLE_NEXT_PROTOCOL_VERSION_UNSAFE_FOR_PRODUCTION
+    LedgerTxn ltxTx(ltx);
+    auto const& resources = sorobanResources();
+    auto lcl = app.getLedgerManager().getLastClosedLedgerNum();
+    auto const& networkConfig =
+        app.getLedgerManager().getSorobanNetworkConfig(ltx);
+
+    auto maxExpirationLedger = lcl + networkConfig.maximumEntryLifetime();
+    auto minRestorableExpirationLedger =
+        lcl + networkConfig.minimumRestorableEntryLifetime();
+    auto minTempExpirationLedger =
+        lcl + networkConfig.minimumTempEntryLifetime();
+    auto autoBumpLedgers = networkConfig.autoBumpLedgers();
+
+    auto autoBumpEnabled = [](LedgerEntry const& le) {
+        return le.data.type() != CONTRACT_DATA ||
+               (le.data.contractData().body.data().flags &
+                ContractDataFlags::NO_AUTOBUMP);
+    };
+
+    auto bump = [&](LedgerEntry& le, auto const& bumpLedgers) {
+        // Use uint64_t do avoid potential overflows from footprint values then
+        // cast down later
+        uint64_t newExpiration = getExpiration(le);
+        if (bumpLedgers)
+        {
+            newExpiration += *bumpLedgers;
+        }
+        else
+        {
+            if (autoBumpEnabled(le))
+            {
+                newExpiration += autoBumpLedgers;
+            }
+        }
+
+        auto minExpirationLedger = isTemporaryEntry(le.data)
+                                       ? minTempExpirationLedger
+                                       : minRestorableExpirationLedger;
+
+        if (newExpiration > maxExpirationLedger)
```

Comment:
Would it make sense to fail at validation step? It doesn't seem hard to avoid specifying too high expiration ledger.